sdk/core/azure-core-amqp/inc/azure/core/amqp/connection_string_credential.hpp

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-Licence-Identifier: MIT
+
+#pragma once
+
+#include "network/sasl_transport.hpp"
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+namespace Azure { namespace Core { namespace _internal { namespace Amqp {
+  //
+  // A ServiceBus connection string has the following format:
+  // "Endpoint=sb://<namespace>.servicebus.windows.net/;SharedAccessKeyName=<KeyName>;SharedAccessKey=<KeyValue>;EntityPath=<entity>"
+  //
+
+  enum class CredentialType
+  {
+    None,
+    SaslPlain,
+    ServiceBusSas,
+    BearerToken,
+  };
+
+  class ConnectionStringCredential {
+  public:
+    ConnectionStringCredential(const std::string& connectionString)
+    {
+      ParseConnectionString(connectionString);
+    }
+    virtual ~ConnectionStringCredential() = default;
+    virtual std::shared_ptr<Network::Transport> GetTransport() const = 0;
+    virtual CredentialType GetCredentialType() const = 0;
```

Comment:
Given that `CredentialType` is countable, make the method non-virtual and take `CredentialType` in constructor?

Reply:
Why add a member variable which does nothing but hold a constant value?

Reply:
If the type can't change after the credential gets created, does not have the complex computation depending on the type, returning a value from a non-virtual method seems more logical to me. This class has relatively many data members, it's not that extra enum member will make a difference.

Reply:
On the other hand, vitual method call is an indirect call, an extra jump, while inline non-virtual getter generates fast and simple code. Not that we worry about perf in that scenario either, but it has its benefits.
Design-wise, I think non-virtual is a more right tool to use in this case (constant behavior, just different value).

Reply:
I ran the experiment on both variants, the codegen is essentially identical. And this is 100% a case of premature optimization - there is no indication that this can ever be on a critical path.

The question to me is: Which implementation is easier to understand - one which initializes a const member in the base with the CredentialType, the other which has a member function that returns a constant value. And honestly, I think it's a wash for readability (although having a const class member variable is not a common pattern).

Show your work: [Godbolt link](https://www.godbolt.org/z/zWjG4zrW8)